src/os/shared.rs

```diff
@@ -34,17 +34,22 @@ impl Iterator for KeyboardEvents {
 
 fn get_datalink_channel(
     interface: &NetworkInterface,
-) -> Result<Box<dyn DataLinkReceiver>, std::io::Error> {
+) -> Result<Box<dyn DataLinkReceiver>, GetInterfaceError> {
     let mut config = Config::default();
     config.read_timeout = Some(time::Duration::new(1, 0));
-
     match datalink::channel(interface, config) {
         Ok(Ethernet(_tx, rx)) => Ok(rx),
-        Ok(_) => Err(std::io::Error::new(
-            ErrorKind::Other,
-            "Unsupported interface type",
-        )),
-        Err(e) => Err(e),
+        Ok(_) => Err(GetInterfaceError::new(GetInterfaceErrorKind::OtherError(
+            "Unsupported interface type".to_string(),
+        ))),
+        Err(e) => match e.kind() {
+            ErrorKind::PermissionDenied => Err(GetInterfaceError::new(
+                GetInterfaceErrorKind::PermissionError(format!("{}. Try running with sudo.", e)),
+            )),
+            _ => Err(GetInterfaceError::new(
```

Comment:
@imsnif  I did not find Result error type which will allow to return here simple std::io::Error `Err(e)`

Reply:
Do you think we might be able to store its text in the GetInterfaceError you created, so that we can display it to the user?

Reply:
@imsnif If user will run `cargo build` without sudo, he will see `Permission error (sudo)`.
If you comment 46-48 lines and run `cargo build` without sudo, you will see `Failed to find any network`, because in this case  we throw  our custom error only in case of Permission Error.

If you want print to user other error `Err(e)` , we should change this [failure::bail!("Failed to find any network interface to listen on.");](https://github.com/imsnif/bandwhich/pull/104/files#diff-79d17550fd953d4eb0860271587ded02L118)

Answer on your question: yes.
Check this [35 line of errors.rs](https://github.com/imsnif/bandwhich/pull/104/files#diff-616fc04220c00148fc9083ac71e33957R35)

Reply:
If you think we should change it, let's do that. :)
The final end goal is:
1. If it's a permission error, the user should get a permission error and have `sudo` suggested to them.
2. If there's another error, let's forward it to the user.
3. If we get several other errors, we can forward all of them to the user. Maybe one per line with the name of the network interface in front of it.

How does that sound?

Reply:
Sounds good. I will do that

Reply:
@imsnif  done.
I made a small refactor to reduce boilerplate

Reply:
Hey, this is great! I'm good to go with this, just one little nitpick:
With this, if there is more than 1 interface that has an error, we will bail and show the error for just the first one.
Would you like to fix this, or shall we leave it for next time?

Reply:
@imsnif I will do it in scope of this PR

Reply:
Awesome! Just another addition: if there is at least one permission error, let's just show the permission message. I think that would be clearer. Sounds good?

Reply:
@imsnif Could you please check ?